projectq/backends/_ionq/_ionq.py

```diff
@@ -160,14 +159,34 @@ def is_available(self, cmd):
     def _reset(self):
         """Reset this backend.
 
+        This also resets the main_engine's qubit allocation index and all
+        allocated qubits.
+
         .. NOTE::
 
             This sets ``_clear = True``, which will trigger state cleanup
             on the next call to ``_store``.
+
+        .. NOTE::
+
+            Because qubit IDs increment without bound and the actual the number
+            of qubits is limited on a per-device basis, the main engine's qubits
+            must be deallocated and the index reset to avoid any key errors or
+            dirty qubit state when the process exits.
         """
-        self._clear = True
-        self._measured_ids = []
+        # Deallocate all previously allocated qubits (this will invoke
+        #   "receive" for each allocated qubit):
+        allocated_qubits = list(self._allocated_qubits)
+        for q in allocated_qubits:
+            self.main_engine.deallocate_qubit(WeakQubitRef(self.main_engine, q))
+        self._allocated_qubits = set()
+
+        # Reset allocation index so we don't end up going past device sizes:
+        self.main_engine._qubit_idx = int(0)
```

Comment:
This is a problem. Hard-coding a reset of the qubit index from the `MainEngine` here deep in some backend code is going to come and bite us back at some point.

Additionally, it is not uncommon to use a backend as a compiler engine (e.g. using a `CommandPrinter` before the simulator). If you arbitrarily reset the internal counter for the qubit IDs here, this has the potential to mess things up.

I would suggest that you remove this and handle the incrementing qubit IDs using a dedicated mapper (which you could automatically add to the `get_engine_list` function in the IonQ setup.

I don't think that we currently have a mapper that is able to do what you want, but I could imaging something along those lines (code not tested):

```python
from projectq.cengines import BasicMapperEngine
from projectq.meta import LogicalQubitIDTag
from projectq.ops import Command, AllocateQubitGate, DeallocateQubitGate
from projectq.types import WeakQubitRef


class BoundedQubitMapper(BasicMapperEngine):
    def __init__(self, n_qubits):
        super().__init__()
        self._qubit_idx = 0
        self.n_qubits = n_qubits

    def _reset(self):
        self._qubit_idx = 0

        # Also automatically send deallocate commands here maybe?

    def _process_cmd(self, cmd):
        if not self.current_mapping:
            self.current_mapping = dict()

        if isinstance(cmd.gate, AllocateQubitGate):
            qubit_id = cmd.qubits[0][0].id
            if qubit_id in self.current_mapping:
                raise RuntimeError("Cannot allocate an already allocated qubit")
            else:
                if self._qubit_idx == n_qubits - 1:
                    raise RuntimeError("Too many qubits")

                new_id = self._qubit_idx
                self._qubit_idx += 1

                self.current_mapping[qubit_id] = new_id

                qb = WeakQubitRef(engine=self, idx=new_id)
                new_cmd = Command(
                    engine=self,
                    gate=AllocateQubitGate(),
                    qubits=([qb],),
                    tags=[LogicalQubitIDTag(qubit_id)],
                )
                self.send([new_cmd])
        elif isinstance(cmd.gate, DeallocateQubitGate):
            qubit_id = cmd.qubits[0][0].id
            if qubit_id not in self.current_mapping:
                raise RuntimeError(
                    "Cannot deallocate a qubit that is not already allocated"
                )
            else:
                qb = WeakQubitRef(engine=self, idx=self.current_mapping[qubit_id])
                new_cmd = Command(
                    engine=self,
                    gate=DeallocateQubitGate(),
                    qubits=([qb],),
                    tags=[LogicalQubitIDTag(qubit_id)],
                )

                self._current_mapping.pop(qubit_id)
                self.send([new_cmd])
        else:
            self._send_cmd_with_mapped_ids(cmd)

    def receive(self, command_list):
        for cmd in command_list:
            if isinstance(cmd.gate, FlushGate):
                self._reset()
                self.send([cmd])
            else:
                self._process_cmd(cmd)
```

Reply:
> This is a problem. Hard-coding a reset of the qubit index from the `MainEngine` here deep in some backend code is going to come and bite us back at some point.
> 
> Additionally, it is not uncommon to use a backend as a compiler engine (e.g. using a `CommandPrinter` before the simulator). If you arbitrarily reset the internal counter for the qubit IDs here, this has the potential to mess things up.
> 
> ```python
> eng = MainEngine(engine_list=[CommandPrinter(), IonQ(...)])
> ```
> 
> I would suggest that you remove this and handle the incrementing qubit IDs using a dedicated mapper (which you could automatically add to the `get_engine_list` function in the IonQ setup.
> 
> I don't think that we currently have a mapper that is able to do what you want, but I could imaging something along those lines (code not tested):
> 
> ```python
> from projectq.cengines import BasicMapperEngine
> from projectq.meta import LogicalQubitIDTag
> from projectq.ops import Command, AllocateQubitGate, DeallocateQubitGate
> from projectq.types import WeakQubitRef
> 
> 
> class BoundedQubitMapper(BasicMapperEngine):
>     def __init__(self, n_qubits):
>         super().__init__()
>         self._qubit_idx = 0
>         self.n_qubits = n_qubits
> 
>     def _reset(self):
>         self._qubit_idx = 0
> 
>         # Also automatically send deallocate commands here maybe?
> 
>     def _process_cmd(self, cmd):
>         if not self.current_mapping:
>             self.current_mapping = dict()
> 
>         if isinstance(cmd.gate, AllocateQubitGate):
>             qubit_id = cmd.qubits[0][0].id
>             if qubit_id in self.current_mapping:
>                 raise RuntimeError("Cannot allocate an already allocated qubit")
>             else:
>                 if self._qubit_idx == n_qubits - 1:
>                     raise RuntimeError("Too many qubits")
> 
>                 new_id = self._qubit_idx
>                 self._qubit_idx += 1
> 
>                 self.current_mapping[qubit_id] = new_id
> 
>                 qb = WeakQubitRef(engine=self, idx=new_id)
>                 new_cmd = Command(
>                     engine=self,
>                     gate=AllocateQubitGate(),
>                     qubits=([qb],),
>                     tags=[LogicalQubitIDTag(qubit_id)],
>                 )
>                 self.send([new_cmd])
>         elif isinstance(cmd.gate, DeallocateQubitGate):
>             qubit_id = cmd.qubits[0][0].id
>             if qubit_id not in self.current_mapping:
>                 raise RuntimeError(
>                     "Cannot deallocate a qubit that is not already allocated"
>                 )
>             else:
>                 qb = WeakQubitRef(engine=self, idx=self.current_mapping[qubit_id])
>                 new_cmd = Command(
>                     engine=self,
>                     gate=DeallocateQubitGate(),
>                     qubits=([qb],),
>                     tags=[LogicalQubitIDTag(qubit_id)],
>                 )
> 
>                 self._current_mapping.pop(qubit_id)
>                 self.send([new_cmd])
>         else:
>             self._send_cmd_with_mapped_ids(cmd)
> 
>     def receive(self, command_list):
>         for cmd in command_list:
>             if isinstance(cmd.gate, FlushGate):
>                 self._reset()
>                 self.send([cmd])
>             else:
>                 self._process_cmd(cmd)
> ```

Ahh, this is the missing piece to the puzzle! I was hoping this would come up. Setting that index felt _really_ awful, but I wasn't sure how else to support this. I did not realize that Mappers were able to process commands like this, so what you have provided here is way better. Thanks for this! 

Will plan to update this today!